### PR TITLE
implement github style link headers

### LINKED ISSUE #224: Linking to headers that contain links
Is it possible, on npm, to link to a header that itself contains a link? For example,

```markdown
## Examples
```

lets you use `#examples` as part of a URL. On the other hand,

```markdown
## [Examples](https://github.com/nwhetsell/csound-api/tree/master/examples)
```

doesn’t let you use `#examples` as part of a URL on npm; https://www.npmjs.com/package/csound-api#examples doesn’t go to the Examples section of csound-api on npm, but https://github.com/nwhetsell/csound-api#examples does on GitHub.

Here’s a short script showing what marky-markdown produces in these cases:

```javascript
const marky = require('marky-markdown');

let HTML = marky('## Examples').html();
console.log(`==> ${HTML}`);
// ==> <h2 id="user-content-examples" class="deep-link"><a href="#examples">Examples</a></h2>

HTML = marky('## [Examples](https://github.com/nwhetsell/csound-api/tree/master/examples)').html();
console.log(`==> ${HTML}`);
// ==> <h2><a href="https://github.com/nwhetsell/csound-api/tree/master/examples">Examples</a></h2>
```

### PR DESCRIPTION
**Pull request by ashleygwilliams**
fixes #224 

to do:

- [x] update tests
- [x] clean up `util`

### CODE REVIEW

File: lib/headings.js [hunk 1]
```diff
@@ -19,22 +19,27 @@ var headings = module.exports = function (md, options) {
...
   md.core.ruler.push('githubHeadings', function (state) {
     // save the Token constructor because we'll be building a few instances at render
     // time; that's sort of outside the intended markdown-it parsing sequence, but
     // since we have tight control over what we're creating (a link), we're safe
     if (!savedTokenConstructor) {
       tokenUtil.set(state.Token)
+      Token = state.Token
       savedTokenConstructor = true
     }
   })
```
Comment (revin) [inline lib/headings.js]: It occurred to me while doing something else that we can eliminate `savedTokenConstructor` from this file and just do
```js
if (!Token) {
  Token = state.Token
  tokenUtil.set(Token)
}
```

File: lib/headings.js [hunk 2]
```diff
@@ -19,22 +19,27 @@ var headings = module.exports = function (md, options) {
...
   md.renderer.rules.heading_open = function (tokens, idx, opts, env, self) {
-    // Bail if heading already contains a hyperlink
+
+    
+
+
     var children = tokens[idx + 1].children
+    // make sure heading is not empty
     if (children && children.length) {
-      if (children.filter(isMarkdownLink).length === 0) {
```
Comment (revin) [inline lib/headings.js]: now we can remove `isMarkdownLink` from up above 🎉

File: lib/headings.js [hunk 3]
```diff
@@ -44,15 +49,21 @@ var headings = module.exports = function (md, options) {
...
+        text.content = "link"
+        var link_close = new Token('link_close', 'a', -1)
+        
+        // add some link attributes
+        link_open.attrSet('id', headings.prefix + postfix)
+        link_open.attrSet('class', headings.className)
+        link_open.attrSet('href', "#" + postfix)
 
-        // if the heading doesn't already contain a link, wrap the contents in one
-        if (children.filter(isHtmlLink).length === 0) {
```
Comment (revin) [inline lib/headings.js]: now we can remove `isHtmlLink` from up above 🎉

File: lib/headings.js [hunk 4]
```diff
@@ -44,15 +49,21 @@ var headings = module.exports = function (md, options) {
...
-          tokenUtil.wrap(children, postfix)
```
Comment (revin) [inline lib/headings.js]: also, so I don't forget, we'll be able to remove the `wrap` function from token-util.js now too.

delete all teh codez

**Comment by ashleygwilliams** [comment]
I LOVE DELETING CODE

**Comment by ashleygwilliams** [comment]
```
119 passing (4s)
  33 failing
```

**Comment by revin** [comment]
33, wow. I guess that part of the rendering is well covered 😛

**Comment by revin** [comment]
✨`152 passing` ✨

**Comment by ashleygwilliams** [comment]
niiiiiiiice 👯

**Comment by ashleygwilliams** [comment]
thought -> should the SVG icon be a flag? @revin

**Comment by revin** [comment]
@ashleygwilliams probably, you're right. We have a slightly now misnamed option `prefixHeadingIds` that enables or disables the 'user-content-' string that gets prepended to the generated `<a>` ids, but nothing for skipping the heading processing completely.

Want an option that disables the link generation altogether? Or an option that enables it, default `true`, and you can set it to false manually if you want? Basically we would be allowing folks to completely skip our githubHeadings plugin completely.

Also since this is already semver-major, should we rename `prefixHeadingIds` to be more accurate? Something like ... `prefixHeadingLinkIds` ...? I don't know, seems long, and you could argue that conceptually the IDs _are_ for the headings, just not the `<h1>`, `<h2>`, etc. elements themselves. I think I just talked myself into leaving that alone.

**Comment by revin** [comment]
Oh wait, did you mean disabling _just_ the icon itself, but still doing the rest of the processing (so we'd still generate the slug and link, but the link would be empty)?

**Comment by ashleygwilliams** [comment]
hah yes, the second thing, just the SVG part. something like `--without-icons`

**Comment by revin** [comment]
I added this as `options.enableHeadingLinkIcons` for programmatic usage; maybe we could add the CLI part in #132?

**Pull request merged**